### ISSUE TITLE
kubernetes/Istio deployment with: telepresence --deployment: upstream connect error or disconnect/reset before headers

### ISSUE THREAD

**Issue by JCzz** [telepresenceio/telepresence#799]
Just to describe the problem:
I can make it work with Istio for 10 seconds with `telepresence --new-deployment debug-deployment-v1 ...`
If I make a kubectl patch deployment debug-deployment-v1 --patch "$(cat kube/debug/patch.yaml)"

```
spec:
  template:
    metadata:
      labels:
        app: debug-deployment
        version: v1
```

But then the container is killed and new one is starting, and I then have to kubectl patch all over again - and unable to succeed in this loop.

I start telepresence with this:

```
telepresence --new-deployment debug-deployment-v1 --docker-run \
  	--rm -it \
	-v $(pwd)/site/build/:/usr/share/nginx/html \
  	nginx bash;
```

The feature request is, to be able to define the deployment used `telepresence --new-deployment`

Update: tried this and not working:
https://www.telepresence.io/reference/connecting /
Running Telepresence manually

Thanks

**Comment by ark3** [comment]
Sorry you're having trouble getting things running. As you surmised, using the `--deployment` option as described in [the docs](https://www.telepresence.io/reference/connecting#running-telepresence-manually) is what I would suggest. Can you describe what is going wrong in that case?

**Comment by JCzz** [comment]
Hi @ark3 

Thanks for reaching out. 
The following is easy to reproduce if you have Istio installed on Kubernetes - it is not hard using Helm on a Google Kubernetes Engine - look [here](https://istio.io/docs/setup/kubernetes/helm-install/#option-2-install-with-helm-and-tiller-via-helm-install) if you dont have Istio installed.

I have tried many things, but as you also suggest - using the Deployment with `datawire/telepresence-k8s:0.93` Note I am also using version 93, not 92 from the doc.

So the steps to reproduce is - first the Deployment, same as on you site for using `datawire/telepresence-k8s:0.93`. I have just added a comment for nginx and opened port 80, that you will use later:

```
Kubectl apply the Deployment:
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: debug-deployment
spec:
  replicas: 1  # only one replica
  template:
    metadata:
      labels:
        name: debug-deployment
        app: debug-deployment
        version: v1
    spec:
      containers:
      - name: debug-deployment
        # image: nginx:latest # Change this
        image: datawire/telepresence-k8s:0.93  # new image
        imagePullPolicy: Always # IfNotPresent
        ports:
        - containerPort: 80
```

Then add a service:
```
# K8s - Service
apiVersion: v1
kind: Service
metadata:
  name: debug-service
  labels:
    app: debug-service
spec:
  ports:
  # Open this port
  - name: http-debug
    port: 6065
    targetPort: 80
  selector:
    app: debug-deployment
```

And finally an Istio Gateway incl. Istio VirtualServics:
```
# Istio
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: debug-gateway
spec:
  selector:
    istio: ingressgateway # use Istio default gateway implementation
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "debug-service.default.svc.cluster.local"
---
# Istio - VirtualService
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: debug-virtualservice
spec:
  hosts:
  - "*"
  gateways:
  - debug-gateway
  http:
  - match:
    - uri:
        prefix: /debug
    rewrite:
      uri: /
    route:
    - destination:
        host: debug-service
        port:
          number: 6065
```

This is important and it is exposing a service/gateway an is assessable via <ip>/debug. `Note the Deployment image is just an nginx server.` And it is replying with the default nginx site.

Now in the deployment under `Change this`, change image to telepresence image, in order to make telepresence work. And then telepresence into the debug-deployment:

And add a site to a sub folder `site/build` You can just use this:
```
<!DOCTYPE html>
<html>
<body>
<h1>This is heading 1</h1>
</body>
</html>
```

And the telepresence:
```
	telepresence --deployment debug-deployment --docker-run \
  	--rm -it \
	-v $(pwd)/site/build/:/usr/share/nginx/html \
	nginx:latest bash;
```
And remember to start nginx with: `/usr/sbin/nginx` And you can verify with a curl localhost in the container - you just need to install curl.

Now if you go and do a `curl <ip>/debug` nothing happens? Well I get:
`upstream connect error or disconnect/reset before headers`

Please notify me on any progress, forwarding or if you have time yourself to look at it - thanks Abhay

**Comment by JCzz** [comment]
Hi @@ark3

Any chance anyone is looking at this?

**Comment by JCzz** [comment]
Hi @ark3

I found the solution,  no bug - the problem was with expose.

Regards

**Comment by ark3** [comment]
I'm glad you were able to get things working. Could you elaborate on how you fixed things? This will help us improve documentation and get other users productive too. Thank you!

**Comment by MartinLG** [comment]
Hi @JCzz 
I am also very interested by you solution on this issue.
Thanks

**Comment by ark3** [comment]
Looking back at this now, I think this might have been #983, which was fixed with Tel 0.99 (just released). Or you can work around the bug by manually adding `--expose 80` or whatever your port is to the command line invocation.
